sw/device/lib/irq.c

```diff
@@ -4,55 +4,47 @@
 
 #include "sw/device/lib/irq.h"
 
-#include "sw/device/lib/base/stdasm.h"
+#include "sw/device/lib/base/csr.h"
 
 static const uint32_t IRQ_EXT_ENABLE_OFFSET = 11;
 static const uint32_t IRQ_TIMER_ENABLE_OFFSET = 7;
 static const uint32_t IRQ_SW_ENABLE_OFFSET = 3;
 
 void irq_set_vector_offset(uintptr_t address) {
-  asm volatile("csrw mtvec, %0" ::"r"(address));
-}
-
-static void irq_mie_set(uint32_t value) {
-  asm volatile("csrrs zero, mie, %0" : : "r"(value) :);
-}
-
-static void irq_mie_clr(uint32_t value) {
-  asm volatile("csrrc zero, mie, %0" : : "r"(value) :);
+  CSR_WRITE(CSR_REG_MTVEC, (uint32_t)address);
 }
 
 void irq_global_ctrl(bool en) {
   if (en) {
-    asm volatile("csrsi mstatus, 0x8" : : :);
+    CSR_SET_BITS(CSR_REG_MSTATUS, 0x8);
   } else {
-    asm volatile("csrci mstatus, 0x8" : : :);
+    CSR_CLEAR_BITS(CSR_REG_MSTATUS, 0x8);
   }
 }
```

Comment:
This honestly makes me wonder if it makes sense to merge SET and CLEAR, with a boolean or enum parameter indicating whether to set or clear. It would certainly reduce places in danger of accidental duplication...

Reply:
It's a good idea but I think I'll leave it out for now. I think there might be times when we want to be explicit about the control flow around CSR operations.